packages/contracts-bedrock/scripts/DeploySuperchain.s.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { Script } from "forge-std/Script.sol";
+
+import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+import { ProtocolVersions, ProtocolVersion } from "src/L1/ProtocolVersions.sol";
+import { ProxyAdmin } from "src/universal/ProxyAdmin.sol";
+import { Proxy } from "src/universal/Proxy.sol";
+
+/// @notice Deploys the Superchain contracts that can be shared by many chains.
+/// We intentionally use the terms "Input" and "Output" to clearly distinguish this script from the
+/// existing ones that use terms of "Config" and "Artifacts".
+contract DeploySuperchain is Script {
+    struct Roles {
+        address proxyAdminOwner;
+        address protocolVersionsOwner;
+        address guardian;
+    }
+
+    struct Input {
+        Roles roles;
+        bool paused;
+        ProtocolVersion requiredProtocolVersion;
+        ProtocolVersion recommendedProtocolVersion;
+    }
+
+    struct Output {
+        ProxyAdmin superchainProxyAdmin;
+        SuperchainConfig superchainConfigImpl;
+        SuperchainConfig superchainConfigProxy;
+        ProtocolVersions protocolVersionsImpl;
+        ProtocolVersions protocolVersionsProxy;
+    }
+
+    /// @notice This entrypoint is for end-users to deploy from an input file and write to an output file.
+    function run(string memory _infile) public returns (Output memory output_, string memory outfile_) {
+        Input memory _input = parseInputFile(_infile);
+        output_ = runWithoutIO(_input);
+        outfile_ = writeOutputFile(_infile, _input, output_);
+        require(false, "DeploySuperchain: run is not implemented");
+    }
+
+    /// @notice This entrypoint is useful for e2e testing purposes, and doesn't use any file I/O.
+    function runWithoutIO(Input memory _input) public returns (Output memory output_) {
```

Comment:
Instead of passing input/output as `memory`, it would be nice to have it structured as:
- a `SuperchainConfig` contract that has the attributes of `Input` (as public getters), and a method to load config from a file.
- a `SuperchainDeployment` contract with setters to write the results to, and a method that writes the final output as file to disk.

And then etch in each contract to some address that is a hash of the contract name, maybe combined with the sender address (so configs/deployments of different tests don't collide).

This would then work well with the Go forge execution, where I can insert precompiles to form the Go config + capture the outputs, without IO, by overriding these input/output contracts with Go.

Any unit-tests could read the addresses from the `SuperchainDeployment` contract, not that different from reading from a contract struct variable.

And, we can then `Semver` the input/output contracts; versioned deploy configuration would be very nice to have.

And the setters/getters of a contract would then all be "type safe" from an ABI standpoint, whereas a memory/calldata struct can have a layout with two swapped address fields and it would be hard to tell apart. The ABI signature here is: `runWithoutIO(((address,address,address),bool,uint256,uint256))`.

But instead it can be one ABI per config variable, and adding/removing configurables and deployment outputs would be much easier to keep compatible (not multiple struct definitions, but rather just additional getters/setters).

And, even better would be to standardize around something like `check()` on the config input contract and the deployment output contract, where we can do some invariant checks (ensure inputs are correct, and ensure outputs are complete/consistent).

Reply:
Took a shot at this alternate approach in https://github.com/ethereum-optimism/optimism/pull/11480 (though it doesn't have `check()` yet), keeping both PRs in draft pending a decision